src/injected/utilityScript.ts

```diff
@@ -0,0 +1,39 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export default class UtilityScript {
+  evaluate(functionText: string, ...args: any[]) {
+    const argCount = args[0] as number;
+    const handleCount = args[argCount + 1] as number;
+    const handles = { __proto__: null } as any;
+    for (let i = 0; i < handleCount; i++)
+      handles[args[argCount + 2 + i]] = args[argCount + 2 + handleCount + i];
+    const visit = (arg: any) => {
+      if ((typeof arg === 'string') && (arg in handles))
+        return handles[arg];
+      if (arg && (typeof arg === 'object')) {
+        for (const name of Object.keys(arg))
+          arg[name] = visit(arg[name]);
+      }
+      return arg;
+    };
+    const processedArgs  = [];
+    for (let i = 0; i < argCount; i++)
+      processedArgs[i] = visit(args[i + 1]);
+    const func = eval('(' + functionText + ')');
```

Comment:
this is going to die with CSP enabled right? Can we inject the function directly into the expression instead of passing functionText?

Reply:
I'm surprised we don't have a test for this by the way, adding one.

Reply:
My bad, we override CSP for our evaluations. Fun!

Then I have a less severe comment:
lets do `window.eval` instead of `eval` so that it doesn't leak all of these local variables into the function scope.